File: DSA/String/Code/RemoveAllOccurrences.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

// Remove All Occurrences of a Character in C++
// Remove all instances of a specific character

int main() {
    string str = "Hello World";
    
    cout << "Original string: " << str << "\n";
    
    // Method 1: Using erase() with remove()
    cout << "\nUsing erase-remove idiom:\n";
    string str1 = str;
    str1.erase(remove(str1.begin(), str1.end(), 'l'), str1.end());
    cout << "After removing all 'l': " << str1 << "\n";
    
    // Method 2: Removing 'o'
    string str2 = str;
    str2.erase(remove(str2.begin(), str2.end(), 'o'), str2.end());
    cout << "After removing all 'o': " << str2 << "\n";
    
    // Method 3: Removing spaces
    string str3 = str;
    str3.erase(remove(str3.begin(), str3.end(), ' '), str3.end());
    cout << "After removing all spaces: " << str3 << "\n";
    
    // Method 4: Using loop and iterator
    cout << "\nUsing loop with iterator:\n";
    string str4 = str;
    for (auto it = str4.begin(); it != str4.end(); ) {
        if (*it == 'l') {
            it = str4.erase(it);
        } else {
            ++it;
        }
    }
    cout << "After removing all 'l': " << str4 << "\n";
    
    // Method 5: Building new string
    cout << "\nBuilding new string:\n";
    string original = "Programming Language";
    string result = "";
    char target = 'a';
    
    for (char c : original) {
        if (c != target) {
            result += c;
        }
    }
    cout << "Original: " << original << "\n";
    cout << "After removing all '" << target << "': " << result << "\n";
    
    return 0;
}

File: DSA/String/Code/RemoveAllOccurrences.py
# Remove All Occurrences of a Character in Python
# Remove all instances of a specific character

str1 = "Hello World"

print("Original string:", str1)

# Method 1: Using replace()
print("\nUsing replace():")
result = str1.replace('l', '')
print("After removing all 'l':", result)

result = str1.replace('o', '')
print("After removing all 'o':", result)

result = str1.replace(' ', '')
print("After removing all spaces:", result)

# Method 2: Using list comprehension
print("\nUsing list comprehension:")
result = ''.join([char for char in str1 if char != 'l'])
print("After removing all 'l':", result)

# Method 3: Using filter()
print("\nUsing filter():")
result = ''.join(filter(lambda x: x != 'l', str1))
print("After removing all 'l':", result)

# Method 4: Using str.translate()
print("\nUsing translate():")
result = str1.translate(str.maketrans('', '', 'l'))
print("After removing all 'l':", result)

# Method 5: Manual approach with loop
print("\nManual approach with loop:")
original = "Programming Language"
target = 'a'
result = ""

for char in original:
    if char != target:
        result += char

print(f"Original: {original}")
print(f"After removing all '{target}': {result}")

# Method 6: Using regex
import re
print("\nUsing regex:")
result = re.sub('l', '', str1)
print("After removing all 'l':", result)

# Method 7: Remove multiple characters
print("\nRemoving multiple characters:")
text = "Hello World!"
result = text.replace('l', '').replace('o', '')
print(f"After removing 'l' and 'o': {result}")

# Method 8: Count removals
print("\nCount before and after:")
original_str = "Mississippi"
target_char = 's'
count_before = original_str.count(target_char)
result = original_str.replace(target_char, '')
count_after = result.count(target_char)
print(f"Original: {original_str}")
print(f"Removed: {count_before} occurrences")
print(f"Result: {result}")
